cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -1316,12 +1335,31 @@ private class ProposalExecutionRunnable implements Runnable {
       _replicationThrottle = replicationThrottle;
       _isTriggeredByUserRequest = isTriggeredByUserRequest;
       _lastSlowTaskReportingTimeMs = -1L;
+      if (removedBrokers != null && !removedBrokers.isEmpty()) {
+        _executionTimerInvolveBrokerRemovalOrDemotion = _proposalExecutionTimerInvolveBrokerRemoval;
+      } else if (demotedBrokers != null && !demotedBrokers.isEmpty()) {
+        _executionTimerInvolveBrokerRemovalOrDemotion = _proposalExecutionTimerInvolveBrokerDemotionOnly;
+      } else {
+        _executionTimerInvolveBrokerRemovalOrDemotion = null;
+      }
     }
 
     public void run() {
       LOG.info("Starting executing balancing proposals.");
-      UserTaskManager.UserTaskInfo userTaskInfo = initExecution();
-      execute(userTaskInfo);
+      final Timer.Context ctxExecutionBrokerRemovalOrDemotion = _executionTimerInvolveBrokerRemovalOrDemotion != null
+          ? _executionTimerInvolveBrokerRemovalOrDemotion.time() : null;
+      final Timer.Context ctxExecution = _proposalExecutionTimer.time();
+      try {
+        UserTaskManager.UserTaskInfo userTaskInfo = initExecution();
+        execute(userTaskInfo);
+      } catch (Exception e) {
+        LOG.error("ProposalExecutionRunnable got exception during run", e);
+      } finally {
+        if (ctxExecutionBrokerRemovalOrDemotion != null) {
+          ctxExecutionBrokerRemovalOrDemotion.stop();
```

Comment:
I'm thinking about another way of implementing it. E.g. have a startTime and an endTime, and then mark each timer with endTime-startTime. 

One potential issue of the current implementation is you need to take care of each individual timer logic here. If you have lots of sensors to mark conditionally, it might be a bit complicated. Eventually we only have a single value to mark, and you can implement the mark timer logic separately regardless how many timers you have. That would decouple the calculate time logic and mark timer logic.

Reply:
Changed to update the timer directly instead of using Context.